election-server/src/bin/http.rs

```diff
@@ -0,0 +1,317 @@
+        env = "CCD_ELECTION_FRONTEND_DIR"
+    )]
+    frontend_dir:         std::path::PathBuf,
+    /// Allow requests from other origins. Useful for development where frontend
+    /// is not served from the server.
+    #[clap(
+        long = "allow-cors",
+        default_value_t = false,
+        env = "CCD_ELECTION_ALLOW_CORS"
+    )]
+    allow_cors:           bool,
+}
+
+/// The app state shared across http requests made to the server.
+#[derive(Clone, Debug)]
+struct AppState {
+    /// The DB connection pool from.
+    db_pool: DatabasePool,
+}
+
+const MAX_SUBMISSIONS_PAGE_SIZE: usize = 20;
+
+fn default_page_size() -> usize { MAX_SUBMISSIONS_PAGE_SIZE }
+
+/// query params passed to [`get_ballot_submissions_by_account`].
+#[derive(Deserialize, Debug)]
+#[serde(rename_all = "camelCase")]
+struct SubmissionsQueryParams {
+    /// The page of ballot submissions to get.
+    #[serde(default)]
+    from:      Option<usize>,
+    /// The pagination size used.
+    #[serde(default = "default_page_size")]
+    page_size: usize,
+}
+
+impl SubmissionsQueryParams {
+    /// Get the page size, where the max page size is capped by
+    /// [`MAX_SUBMISSIONS_PAGE_SIZE`]
+    fn page_size(&self) -> usize { cmp::min(self.page_size, MAX_SUBMISSIONS_PAGE_SIZE) }
+}
+
+/// The response type for [`get_ballot_submissions_by_account`] queries
+#[derive(Serialize)]
+#[serde(rename_all = "camelCase")]
+struct SubmissionsResponse {
+    /// Ballots returned in the response
+    results:  Vec<StoredBallotSubmission>,
+    /// Whether there are more results in the database
+    has_more: bool,
+}
+
+/// Get ballot submissions registered for `account_address`. Returns
+/// [`StatusCode`] signaling error if database connection or lookup fails.
+#[tracing::instrument(skip(state))]
+async fn get_ballot_submissions_by_account(
+    State(state): State<AppState>,
+    Path(account_address): Path<AccountAddress>,
+    Query(query_params): Query<SubmissionsQueryParams>,
+) -> Result<Json<SubmissionsResponse>, StatusCode> {
+    let db = state.db_pool.get().await.map_err(|e| {
+        tracing::error!("Could not get db connection from pool: {}", e);
+        StatusCode::INTERNAL_SERVER_ERROR
+    })?;
+
+    let page_size = query_params.page_size();
+    let mut results = db
+        // Add 1 to the page size to identify if there are more results on the next "page"
+        .get_ballot_submissions(account_address, query_params.from, page_size + 1)
+        .await
+        .map_err(|e| {
+            tracing::error!("Failed to get ballot submissions for account: {}", e);
+            StatusCode::INTERNAL_SERVER_ERROR
+        })?;
+
+    let has_more = results.len() > query_params.page_size();
+    if has_more {
+        // Pop the first value of the result, as the query returns values in descending
+        // order
+        let mut results_deque: VecDeque<_> = results.into();
+        results_deque.pop_front();
+        results = results_deque.into();
```

Comment:
I could be confused, but isn't this wrong? You are returning values from "most recent" to "oldest", right?

So should you not be omitting the last value here?

Reply:
Haha, this ordering business has confused me more than I'm proud to admit 😆 I would think you're right in your assumption though 😄